yktrace/src/hwt.rs

```diff
@@ -0,0 +1,94 @@
+//! Hardware tracing via ykrustc.
+
+use super::{SirTrace, ThreadTracer, ThreadTracerImpl};
+use crate::{errors::InvalidTraceError, maptrace::MapTrace};
+use core::yk::SirLoc;
+use hwtracer::backends::TracerBuilder;
+
+/// A trace collected via hardware tracing.
+#[derive(Debug)]
+struct HWTSirTrace {
+    sirtrace: Vec<SirLoc>
+}
+
+impl SirTrace for HWTSirTrace {
+    fn raw_len(&self) -> usize {
+        self.sirtrace.len()
+    }
+
+    fn raw_loc(&self, idx: usize) -> &SirLoc {
+        &self.sirtrace[idx]
+    }
+}
+
+/// Hardware thread tracer.
+struct HWTThreadTracer {
+    ttracer: Box<dyn hwtracer::ThreadTracer>
+}
+
+impl ThreadTracerImpl for HWTThreadTracer {
+    #[trace_tail]
+    fn stop_tracing(&mut self) -> Result<Box<dyn SirTrace>, InvalidTraceError> {
+        let hwtrace = self.ttracer.stop_tracing().unwrap();
+        let mt = MapTrace::new();
+        if let Some(sirtrace) = mt.map(hwtrace) {
+            Ok(Box::new(HWTSirTrace { sirtrace }) as Box<dyn SirTrace>)
+        } else {
+            Err(InvalidTraceError::InternalError)
+        }
+    }
+}
+
+#[trace_head]
+pub fn start_tracing() -> ThreadTracer {
+    let tracer = TracerBuilder::new().build().unwrap();
+    let mut ttracer = (*tracer).thread_tracer();
+    match ttracer.start_tracing() {
+        Ok(()) => {}
+        Err(e) => panic!("Failed to start tracer: {}", e)
+    }
+
+    ThreadTracer {
+        // FIXME double boxed value
```

Comment:
Is that still the case?

Reply:
Yes, `ttracer` is a `Box<dyn hwtracer::ThreadTracer>`. Preventing the double boxing would require `HWTThreadTracer` be a dst. While that it possible, there is no safe way to construct it in that case.

```rust
#[repr(transparent)]
pub struct A {
    a: dyn std::error::Error,
}

pub fn make_a(a: Box<dyn std::error::Error>) -> Box<A> {
    // I think this is safe, because `#[repr(transparent)]` guarantees that the layout is of `A` is the same as `dyn std::error::Error`.
    unsafe { std::mem::transmute(a) }
}
```

Reply:
dst?

Reply:
A dst is a dynamically sized type, for example `[T]`, or `dyn Trait`. Tese types don't have a known size at compile time.

Reply:
Ok, if it's unfixable, I will just remove the comment.